plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -849,13 +846,17 @@ void producer_plugin::plugin_startup()
    ilog("producer plugin:  plugin_startup() begin");
 
    chain::controller& chain = my->chain_plug->chain();
-   my->_unapplied_transactions.set_chain_read_mode( chain.get_read_mode() );
    EOS_ASSERT( my->_producers.empty() || chain.get_read_mode() == chain::db_read_mode::SPECULATIVE, plugin_config_exception,
               "node cannot have any producer-name configured because block production is impossible when read_mode is not \"speculative\"" );
 
    EOS_ASSERT( my->_producers.empty() || chain.get_validation_mode() == chain::validation_mode::FULL, plugin_config_exception,
               "node cannot have any producer-name configured because block production is not safe when validation_mode is not \"full\"" );
 
+   unapplied_transaction_queue::process_mode unapplied_mode =
+         (chain.get_read_mode() != chain::db_read_mode::SPECULATIVE) ? unapplied_transaction_queue::process_mode::non_speculative :
+            my->_producers.empty() ? unapplied_transaction_queue::process_mode::speculative_non_producer :
+               unapplied_transaction_queue::process_mode::speculative_producer;
+   my->_unapplied_transactions.set_mode( unapplied_mode );
```

Comment:
Do this in `producer_plugin::plugin_initiailize`. Since producer_plugin depends on chain_plugin, it is safe to get the read_mode from the controller at that point.